include/transactions.hpp

```diff
@@ -0,0 +1,497 @@
+
+    auto
+    operator<=>(const TransactionResult &) const noexcept
+            = default;
+};
+
+template<typename T, typename parsePred, typename matchPred>
+struct CtxResult {
+    const TimingCtx<parsePred, matchPred> timingCtx;
+    const settings::node<T>              &settingValue;
+
+    explicit(false) constexpr             operator auto const &() const noexcept { return timingCtx; }
+
+    explicit(false) constexpr             operator T const &() const noexcept { return settingValue; }
+
+    auto
+    operator<=>(const CtxResult &) const noexcept
+            = default;
+};
+
+} // namespace settings
+
+template<std::equality_comparable TransactionToken>
+inline static const TransactionToken NullToken = TransactionToken{};
+
+/**
+ * @brief A thread-safe settings wrapper that supports multiple stage/commit transactions and history functionality.
+ *
+ * Example:
+ * @code
+ * fair::graph::SettingBase<int,int, std::string, 16, std::chrono::seconds, 3600 * 24, 10> settings;
+ *
+ * auto [ok1, timeStamp1] = settings.commit(42); // store 42 in settings
+ * auto [ok2, timeStamp2] = settings.commit(43); // store 43 in settings
+ *
+ * assert(settings.get() == 43); // get the latest value
+ * assert(settings.get(timeStamp2) == 43); // get the first isCommitted value since timeStamp2
+ * assert(settings.get(-1) == 42); // get the second to last value (for lib development only)
+ * assert(settings.get(timeStamp1) == 42); // get the first isCommitted value since timeStamp1
+ *
+ * auto [ok3, timeStamp3] = settings.stage(53, "transactionToken#1"); // stage 53 in settings, N.B. setting is not yet committed
+ * assert(settings.get() == 43); // get the latest value
+ * auto [ok4, timeStamp4] = settings.commit("transactionToken#1"); // commit transaction
+ * assert(settings.get() == 53); // get the latest value
+ * @endcode
+ *
+ * @tparam T is the user-supplied setting type, for simple settings U and T are identical (see CtxSettings for an example where it isn't).
+ * @tparam U is the internally stored setting type that may include additional meta data.
+ * @tparam TransactionToken unique identifier with which to store/commit transactions.
+ * @tparam N_HISTORY the maximum number of setting history
+ * @tparam TimeDiff the std::chrono::duration time-base for the time-outs
+ * @tparam timeOut maximum time given in units of TimeDiff after which a setting automatically expires if unused. (default: -1 -> disabled)
+ * @tparam timeOutTransactions maximum time given in units of TimeDiff after which a transaction automatically expires if not being committed. (default: -1 -> disabled)
+ */
+template<std::movable T, std::movable U = T, std::equality_comparable TransactionToken = std::string, std::size_t N_HISTORY = 1024, typename TimeDiff = std::chrono::seconds, int timeOut = -1,
+         int timeOutTransactions = -1>
+    requires(fair::meta::is_power2_v<N_HISTORY> && N_HISTORY > 8)
```

Comment:
Is this still functionally needed or just a performance optimisation?

Reply:
I transfered this requirement from the opencmw implementation, I guess its purpose was just performance.